src/devices/cpu/drcbex64.cpp

```diff
@@ -3332,160 +3568,195 @@ void drcbe_x64::op_muls(Assembler &a, const instruction &inst)
+	else if (inst.size() == 8)
+	{
+		Gp dstreg = rax;
+		Gp hireg = rdx;
+
+		mov_reg_param(a, dstreg, src1p);
+		if (src2p.is_memory())
+			a.imul(MABS(src2p.memory(), 8));
+		else if (src2p.is_int_register())
+			a.imul(Gpq(src2p.ireg()));
+		else if (src2p.is_immediate())
 		{
-			mov_reg_param(a, rax, src1p);                                               // mov   rax,src1p
-			if (src2p.is_memory())
-				a.imul(MABS(src2p.memory(), 8));                                        // imul  [src2p]
-			else if (src2p.is_int_register())
-				a.imul(Gpq(src2p.ireg()));                                              // imul  src2p
-			else if (src2p.is_immediate())
-			{
-				mov_r64_imm(a, rdx, src2p.immediate());                                 // mov   rdx,src2p
-				a.imul(rdx);                                                            // imul  rdx
-			}
-			mov_param_reg(a, dstp, rax);                                                // mov   dstp,rax
-			mov_param_reg(a, edstp, rdx);                                               // mov   edstp,rdx
+			mov_r64_imm(a, hireg, src2p.immediate());
+			a.imul(hireg);
 		}
+		mov_param_reg(a, dstp, dstreg);
 
 		// compute flags
-		if (inst.flags() != 0)
+		if (inst.flags())
 		{
-			if (zsflags != 0)
-			{
-				if (vflag)
-					a.pushfq();                                                         // pushf
-				if (compute_hi)
-				{
-					if (zsflags == FLAG_Z)
-						a.or_(rdx, rax);                                                // or    rdx,rax
-					else if (zsflags == FLAG_S)
-						a.test(rdx, rdx);                                               // test  rdx,rdx
-					else
-					{
-						a.mov(ecx, eax);                                                // mov   ecx,eax
-						a.shr(rax, 32);                                                 // shr   rax,32
-						a.or_(rdx, rcx);                                                // or    rdx,rcx
-						a.or_(rdx, rax);                                                // or    rdx,rax
-					}
-				}
-				else
-					a.test(rax, rax);                                                   // test  rax,rax
+			a.test(dstreg, dstreg);
+			a.pushfd(); // sign + zero
 
-				// we rely on the fact that OF is cleared by all logical operations above
-				if (vflag)
-				{
-					a.pushfq();                                                         // pushf
-					a.pop(rax);                                                         // pop   rax
-					a.and_(qword_ptr(rsp), ~0x84);                                      // and   [rsp],~0x84
-					a.or_(ptr(rsp), rax);                                               // or    [rsp],rax
-					a.popfq();                                                          // popf
-				}
-			}
+			a.mov(rcx, hireg);
+			a.cqo();
+
+			a.cmp(rcx, hireg);
+			a.pushfd();
+			a.pop(hireg);
+			a.and_(hireg, 0x40); // zero
+			a.xor_(hireg, 0x40);
+			a.shl(hireg, 5); // turn into overflow flag
+			a.or_(dword_ptr(rsp, 0), hireg);
+
+			a.popfd();
```

Comment:
There are potential issues with storing flags on the stack:
* x86-64 ABIs expect the stack pointer to be static and remain 16-byte aligned outside function prologues and epilogues.  Violating this can cause problems with signal handler and interrupts.
* `pushfd` and `popfd` tend to cause pipeline stalls and hurt performance.

Can you use `lahf` and `sahf` instead?

Reply:
https://github.com/mamedev/mame/pull/13108/commits/25b8ab97d96e07b4569702bf90b072ee4164d5b5

I used r10 and r11 as scratch registers to handle juggling register values. I [double checked](https://learn.microsoft.com/en-us/cpp/build/x64-software-conventions?view=msvc-170#x64-register-usage) and for Windows r10 and r11 considered volatile so I think it should be safe to use them. The header of the x64 DRC also says for Linux/MacOS the r10 and r11 registers are volatile.

x86 uses pushfd/popfd a lot more and there aren't really any readily available scratch registers to make it easier to implement so I don't think it's worth the time to optimize that for x86. Someone else can handle that if they think it's worth the effort though.